### PR TITLE
Add new column for proxy count

### PR DESCRIPTION
**Pull request by JeremieSamson**
A procuration proxy can now have more than one request associated (OneToOne => OneToMany) based on those two rules :
- A French proxy (voting in France) can receive max one request from France or two requests made in another country or one of each.
- A foreign proxy (voting in another country) can receive one request from France, or one in France and one (or two) from another country, than France or 3 from another chountry than France. 

_You will have all details here (in french) =>_ https://www.service-public.fr/particuliers/vosdroits/F1604

Severals changes on the front

**First, on the new proxy form (`procuration/je-propose`)**

fr

non fr

_from #3415 the only thing that changed is the rule saying that a foreign proxy can take 3 requests from another country._

**Second, the request form (`procuration/je-demande/ma-procuration`)**

where

_Here the user have to say where he made his request (in France or in another country)_

Some other change have been made on the back side

**First, relation one to one / one to many**

```php
/**
     * The associated found request(s).
     *
     * @var ArrayCollection
     *
     * @ORM\OneToMany(targetEntity="AppBundle\Entity\ProcurationRequest", mappedBy="foundProxy")
     */
    private $foundRequests;
```

A proxy can now have multiple requests
A request can only have one proxy

```php
/**
     * The associated found proxy.
     *
     * @var ProcurationProxy
     *
     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\ProcurationProxy", inversedBy="foundRequests")
     */
    private $foundProxy;
```

**Then, on the `ProcurationProxy` entity, booleans have been added to apply the new rule**

```php
/**
     * @var bool
     *
     * @ORM\Column(type="boolean", options={"default": true})
     */
    public $frenchRequestAvailable = true;

    /**
     * @var bool
     *
     * @ORM\Column(type="boolean", options={"default": true})
     */
    public $foreignRequestAvailable = true;
```

those two booleans will be drive by the `process` and `unprocess` functions called when the procuration manager will associate or disassociate a proxy and a request.
I choose boolean because I can centralize the new rules in one place and use them in the repository to ease proxy responsable space navigation.

**Moreover we have the rules that apply this feature in the function**

```php
public function process(ProcurationProxy $procurationProxy = null, Adherent $procurationFoundBy = null): void
    {
        $this->foundProxy = $procurationProxy;
        $this->foundBy = $procurationFoundBy;
        $this->processed = true;
        $this->processedAt = new \DateTimeImmutable();

        if ($procurationProxy) {
            $proxiesUsedCount = $procurationProxy->getFoundRequests()->count();
            $remainingProxiesCount = $procurationProxy->proxiesCount - $proxiesUsedCount;

            if (1 === $remainingProxiesCount) {
                $procurationProxy->setFrenchRequestAvailable(false);
                $procurationProxy->setForeignRequestAvailable(false);
            } else {
                if ('FR' === $procurationProxy->getVoteCountry()) {
                    $procurationProxy->setFrenchRequestAvailable(false);
                } else {
                    if (2 === $remainingProxiesCount && !$procurationProxy->isFrenchRequestAvailable()) {
                        $procurationProxy->setForeignRequestAvailable(false);
                    }
                }
            }

            $procurationProxy->addFoundRequest($this);
        }
    }

public function unprocess(): void
    {
        if ($this->foundProxy instanceof ProcurationProxy) {
            $proxiesUsedCount = $this->foundProxy->getFoundRequests()->count();
            $remainingProxiesCount = $this->foundProxy->proxiesCount - $proxiesUsedCount;

            if (1 === $this->foundProxy->getFoundRequests()->count()) {
                $this->foundProxy->setFrenchRequestAvailable(true);
                $this->foundProxy->setForeignRequestAvailable(true);
            } else {
                if ('FR' === $this->foundProxy->getVoteCountry()) {
                    $this->foundProxy->setFrenchRequestAvailable(true);
                } else {
                    $this->foundProxy->setForeignRequestAvailable(true);

                    if (0 === $remainingProxiesCount) {
                        $this->foundProxy->setFrenchRequestAvailable(true);
                    }
                }
            }

            $this->foundProxy->removeFoundRequest($this);
        }

        $this->foundProxy = null;
        $this->processed = false;
        $this->processedAt = null;
    }
```

### CODE REVIEW

File: features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -83,4 +83,16 @@ public function assertPageAddressAfterAllRedirection(string $url, int $ttl = 3,
             $this->assertPageAddressAfterAllRedirection($url, $ttl, ++$try);
         }
     }
+
+    /**
+     * @When I click the :cssElementSelector
+     */
+    public function clickElement($cssElementSelector)
```
Comment (JeremieSamson) [inline features/bootstrap/FeatureContext.php]: this new step is to force clicking on an element directly when it's not working with a basic check with the @javascript tag on

(thanks @ottaviano )

File: src/Entity/ProcurationRequest.php [hunk 2]
```diff
@@ -344,14 +351,48 @@ public function process(ProcurationProxy $procurationProxy = null, Adherent $pro
         $this->processedAt = new \DateTimeImmutable();
 
         if ($procurationProxy) {
-            $procurationProxy->setFoundRequest($this);
+            $proxiesUsedCount = $procurationProxy->getFoundRequests()->count();
```
Comment (JeremieSamson) [inline src/Entity/ProcurationRequest.php]: I'm pretty sure those two logics (from process and unprocess) could be improved, open to discussions :)

File: src/Entity/ProcurationRequest.php [hunk 3]
```diff
@@ -310,6 +310,13 @@ class ProcurationRequest
      */
     public $recaptcha = '';
 
+    /**
+     * @var bool
+     *
+     * @ORM\Column(type="boolean", options={"default": true})
+     */
+    private $requestFromFrance = true;
```
Comment (JeremieSamson) [inline src/Entity/ProcurationRequest.php]: New param to know if proxy has been made in France or not

File: src/Procuration/Filter/ProcurationProxyProposalFilters.php [hunk 4]
```diff
@@ -36,14 +36,28 @@ public function apply(QueryBuilder $qb, string $alias): void
...
+                    $qb->expr()->orX(
+                        "$alias.frenchRequestAvailable != 0",
+                        "$alias.foreignRequestAvailable != 0"
+                    )
+                )
+                ->setParameter('disabled', false)
+             ;
         } elseif (self::ASSOCIATED === $status) {
-            $qb->andWhere("$alias.foundRequest IS NOT NULL");
+            $qb
```
Comment (JeremieSamson) [inline src/Procuration/Filter/ProcurationProxyProposalFilters.php]: PO sayed associated will no longer be a simple associated proxy but a fully associated proxy (no more availability)

File: templates/procuration/proposal.html.twig [hunk 5]
```diff
@@ -45,11 +45,11 @@
 
                 if ('FR' === $('#{{ procuration_form.voteCountry.vars.id }}').val()) {
                     $(proxiesCountSelected+'_'+1).prop('checked', true);
-                    radioButtons.last().hide();
+                    radioButtons.last().addClass("hidden");
```
Comment (JeremieSamson) [inline templates/procuration/proposal.html.twig]: hidden class is better for test purpose

File: tests/Entity/ProcurationRequestTest.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        $request->process($proxy);
+
+        $this->assertEquals(1, $proxy->getFoundRequests()->count());
+        $this->assertEquals(true, $proxy->isFrenchRequestAvailable());
+        $this->assertEquals(true, $proxy->isForeignRequestAvailable());
+    }
+
+    public function testProcessAndUnProcessWithOneProxyCount()
+    {
+        $proxy = new ProcurationProxy(null);
```
Comment (thomasbisignani) [inline tests/Entity/ProcurationRequestTest.php]: Just a first pre-review comment, for your all tests, don't hesitate to use a DataProvider to refactor and use easly the expected values

File: tests/Entity/ProcurationRequestTest.php [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        $proxy->setForeignRequestAvailable(true);
+
+        $request = new ProcurationRequest();
+        $request->setRequestFromFrance(0);
+
+        $request->process($proxy);
+
+        $this->assertEquals(1, $proxy->getFoundRequests()->count());
```
Comment (thomasbisignani) [inline tests/Entity/ProcurationRequestTest.php]: Just a first pre-review comment, for your all tests, don't hesitate to use a DataProvider to refactor and use easly the expected values

Comment (JeremieSamson) [reply tests/Entity/ProcurationRequestTest.php]: let me know what you think of the last commit @thomasbisignani ;)

File: src/Repository/ProcurationProxyRepository.php [hunk 8]
```diff
@@ -104,7 +106,6 @@ public function findMatchingProxies(ProcurationRequest $procurationRequest)
                     'pp.voteCountry = :voteCountry'
                 )
             ))
-            ->andWhere('pp.foundRequest IS NULL')
```
Comment (ottaviano) [inline src/Repository/ProcurationProxyRepository.php]: do you need to replace this check by something like 
```sql
(country = 'FR' AND pp.frenchRequestAvailable)
OR (country != 'FR' AND pp.foreignRequestAvailable)
```

File: templates/procuration_manager/_proposals_list.html.twig [hunk 9]
```diff
@@ -27,18 +27,22 @@
             </span>
         </td>
         <td class="datagrid__table__col--left">
-            {{ include('procuration_manager/_matching_election_rounds.html.twig', {
-                proxy_rounds: proxy.availableRounds,
-                request: proxy.foundRequest,
-            }, with_context=false) }}
+            {{ proxy.getAvailableRoundsAsString|raw }}
```
Comment (ottaviano) [inline templates/procuration_manager/_proposals_list.html.twig]: you can keep `implode("\n", ...)` on the methode and use here `|nl2br` twig filter

File: templates/procuration_manager/associate.html.twig [hunk 10]
```diff
@@ -148,10 +148,7 @@
             </div>
             <div class="profile-value">
                 <span id="proxy-election">
-                    {{ include('procuration_manager/_matching_election_rounds.html.twig', {
-                        proxy_rounds: proxy.electionRounds,
-                        request: request
-                    }, with_context=false) }}
+                    {{ proxy.getAvailableRoundsAsString|raw }}}
```
Comment (ottaviano) [inline templates/procuration_manager/associate.html.twig]: same here

File: templates/procuration_manager/deassociate.html.twig [hunk 11]
```diff
@@ -151,10 +151,7 @@
                 Disponibilités
             </div>
             <div class="profile-value">
-                {{ include('procuration_manager/_matching_election_rounds.html.twig', {
-                    proxy_rounds: proxy.availableRounds,
-                    request: proxy.foundRequest,
-                }, with_context=false) }}
+                {{ proxy.getAvailableRoundsAsString|raw }}
```
Comment (ottaviano) [inline templates/procuration_manager/deassociate.html.twig]: same

File: app/migrations/Version20190227153320.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                'procuration_proxy_id' => $procurationProxy['procuration_proxy_id'],
+                'procuration_request_id' => $procurationProxy['procuration_request_id'],
+                'vote_country' => $procurationProxy['vote_country'],
+            ];
+        }
+    }
+
+    public function up(Schema $schema): void
+    {
+        $this->addSql('UPDATE procuration_requests SET procuration_request_found_by_id = NULL WHERE procuration_request_found_by_id IS NOT NULL');
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply app/migrations/Version20190227153320.php]: This should be an old behaviour, nice catch

File: features/bootstrap/FeatureContext.php [hunk 13]
```diff
@@ -83,4 +83,16 @@ public function assertPageAddressAfterAllRedirection(string $url, int $ttl = 3,
             $this->assertPageAddressAfterAllRedirection($url, $ttl, ++$try);
         }
     }
+
+    /**
+     * @When I click the :cssElementSelector
```
Comment (thomasbisignani) [inline features/bootstrap/FeatureContext.php]: What about adding more precision like :
 ```php
I click the :cssElementSelector selector
```

We will have two clearly distinct cases : 

`I click the "toto" selector` for CSS selectors
`I click the "toto" element` for ID elements

File: src/Repository/ProcurationProxyRepository.php (hunk 8 above)
Comment (JeremieSamson) [reply src/Repository/ProcurationProxyRepository.php]: yes, should be similar query that the one who count the number of match :+1:

File: features/bootstrap/FeatureContext.php (hunk 13 above)
Comment (JeremieSamson) [reply features/bootstrap/FeatureContext.php]: yep good idea :+1:

File: tests/Entity/ProcurationProxyTest.php [hunk 14]
```diff
@@ -10,22 +10,69 @@
...
+        $proxy->setProxiesCount($proxiesCount);
+        $proxy->addFoundRequest($request);
+        $proxy->setElectionRounds(\array_slice($electionRounds, 0, $proxyElectionRoundsCount));
+
+        $this->assertCount($expectedAvailableElectionRoundsCount, $available = $proxy->getAvailableRounds());
+    }
+
+    public function provideTestCases(): \Generator
+    {
+        yield [1, 1, 1, 0];
```
Comment (thomasbisignani) [inline tests/Entity/ProcurationProxyTest.php]: Can you just add comments by "block" to explain in few words ? 

Or maybe you can add a description by `yield` like : 

```php
yield 'Your beautiful description' => [1, 1, 2, 1];
```

File: src/Entity/ProcurationProxy.php [hunk 15]
```diff
@@ -560,17 +572,31 @@ public function setVoteOffice(string $voteOffice): void
     }
 
     /**
-     * @return ElectionRound[]|Collection
+     * @return ElectionRound[]|Collection|null
      */
-    public function getAvailableRounds(): Collection
+    public function getAvailableRounds(): ?Collection
```
Comment (ottaviano) [inline src/Entity/ProcurationProxy.php]: when can it be null?

File: src/Entity/ProcurationProxy.php [hunk 16]
```diff
@@ -585,17 +611,25 @@ public function getAvailableRoundsAsString(): string
             return self::NO_AVAILABLE_ROUND;
         }
 
-        return implode("\n", $availableRounds->toArray());
+        return implode('\n', $availableRounds->toArray());
```
Comment (ottaviano) [inline src/Entity/ProcurationProxy.php]: it should be `"` instead of `'`, no?

File: src/Entity/ProcurationProxy.php [hunk 17]
```diff
@@ -585,17 +611,25 @@ public function getAvailableRoundsAsString(): string
...
+    }
+
+    public function getFoundRequests(): Collection
+    {
+        return $this->foundRequests;
     }
 
-    public function getFoundRequest(): ?ProcurationRequest
+    public function addFoundRequest(ProcurationRequest $procurationRequest): void
```
Comment (ottaviano) [inline src/Entity/ProcurationProxy.php]: why do we need adder and remover methods? 
we control this with `process` and `unprocess` methods no?

File: src/Entity/ProcurationProxy.php [hunk 18]
```diff
@@ -666,4 +700,68 @@ public function setProxiesCount(int $proxiesCount): void
     {
         $this->proxiesCount = $proxiesCount;
     }
+
+    public function isFrenchRequestAvailable(): bool
+    {
+        return $this->frenchRequestAvailable;
+    }
+
+    public function setFrenchRequestAvailable(bool $frenchRequestAvailable): void
```
Comment (ottaviano) [inline src/Entity/ProcurationProxy.php]: IMO I think we don't need to open these control methods, just process and unprocess are public for change this state, WDYT?

File: src/Form/Procuration/ProcurationProxyType.php [hunk 19]
```diff
@@ -46,6 +46,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('voteOffice', TextType::class)
             ->add('proxiesCount', ChoiceType::class, [
                 'expanded' => true,
+                'mapped' => false,
```
Comment (ottaviano) [inline src/Form/Procuration/ProcurationProxyType.php]: why this change? should it not be mapped to `Proxy::proxiesCount`?

File: src/Entity/ProcurationProxy.php (hunk 17 above)
Comment (JeremieSamson) [reply src/Entity/ProcurationProxy.php]: the `process()` and `unprocess()` function will use those functions (`addFoundRequest()` & `removeFoundRequest`) right after checking if it's possible or not

File: src/Entity/ProcurationProxy.php (hunk 18 above)
Comment (JeremieSamson) [reply src/Entity/ProcurationProxy.php]: yep good idea :+1:

File: src/Entity/ProcurationProxy.php (hunk 17 above)
Comment (ottaviano) [reply src/Entity/ProcurationProxy.php]: so they can be private to avoid using it?

File: src/Form/Procuration/ProcurationProxyType.php (hunk 19 above)
Comment (JeremieSamson) [reply src/Form/Procuration/ProcurationProxyType.php]: nice catch

File: features/bootstrap/FeatureContext.php [hunk 20]
```diff
@@ -83,4 +83,16 @@ public function assertPageAddressAfterAllRedirection(string $url, int $ttl = 3,
...
     }
+
+    /**
+     * @When I click the :cssElementSelector selector
+     */
+    public function clickElementSelector($cssElementSelector)
+    {
+        $field = $this->getSession()->getPage()->find('css', $cssElementSelector);
+
+        Assert::notNull($field, 'Cannot find "'.$cssElementSelector.'"');
```
Comment (thomasbisignani) [inline features/bootstrap/FeatureContext.php]: Not a big deal, but you can write easier  : 

```php
 Assert::notNull($field, "Cannot find '$cssElementSelector'");
```

File: tests/Entity/ProcurationRequestTest.php [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+        $request->process($proxy);
+
+        $this->assertEquals($isFrenchRequestAvailable, $proxy->isFrenchRequestAvailable());
+        $this->assertEquals($isForeignRequestAvailable, $proxy->isForeignRequestAvailable());
+    }
+
+    public function provideTestCases(): \Generator
+    {
+        // For one request max by procuration proxy
+        yield 'An associated french proxy with a french request made in France should result with no more availability' => [1, 'FR', 1, 'FR', false, false];
```
Comment (thomasbisignani) [inline tests/Entity/ProcurationRequestTest.php]: CS : Can you multiple line here please ?

File: tests/Entity/ProcurationProxyTest.php [hunk 22]
```diff
@@ -10,22 +10,72 @@
...
+        $proxy->process($request);
+        $proxy->setElectionRounds(\array_slice($electionRounds, 0, $proxyElectionRoundsCount));
+
+        $this->assertCount($expectedAvailableElectionRoundsCount, $available = $proxy->getAvailableRounds());
+    }
+
+    public function provideTestCases(): \Generator
+    {
+        // For one request max by procuration proxy
+        yield 'One election for the request and one for the proxy should not give available election' => [1, 1, 1, 0];
```
Comment (thomasbisignani) [inline tests/Entity/ProcurationProxyTest.php]: CS : Can you multiple line here ?

File: features/bootstrap/FeatureContext.php (hunk 20 above)
Comment (JeremieSamson) [reply features/bootstrap/FeatureContext.php]: I guess the `''` are here to isolate the `$cssElementSelector` in the message :thinking:
